src/persist-client/src/write.rs

```diff
@@ -266,36 +294,72 @@ where
             new_upper
         );
 
+        let batch = match self
+            .write_batch(updates, expected_upper.clone(), new_upper.clone())
+            .await
+        {
+            Ok(batch) => batch,
+            Err(invalid_usage) => return Ok(Err(invalid_usage)),
+        };
+
+        self.compare_and_append_core(batch, expected_upper, new_upper)
+            .await
+    }
+
+    /// Appends the given batch of updates to this shard and downgrades this
+    /// handle's upper to `new_upper` iff the current global upper of this shard
+    /// is `expected_upper`.
+    ///
+    /// The innermost `Result` is `Ok` if the batch was successfully written. If
+    /// not, an `Upper` err containing the current global upper is returned.
+    ///
+    /// In contrast to [Self::append_core], this linearizes mutations from all
+    /// writers. It's intended for use as an atomic primitive for timestamp
+    /// bindings, SQL tables, etc.
+    ///
+    /// A `new_upper` of the empty antichain "finishes" this shard, promising
+    /// that no more data is ever incoming.
+    ///
+    /// The batch may be empty, which allows for downgrading `upper` to
+    /// communicate progress. It is possible to heartbeat a writer lease by
+    /// calling this with `new_upper` equal to `self.upper()` and an empty
+    /// `updates` (making the call a no-op).
+    ///
+    /// The clunky multi-level Result is to enable more obvious error handling
+    /// in the caller. See <http://sled.rs/errors.html> for details.
+    ///
+    /// SUBTLE! Unlike the other methods on WriteHandle, it is not always safe
+    /// to retry [ExternalError]s in compare_and_append (depends on the usage
+    /// pattern). We should be able to structure timestamp binding, source, and
+    /// sink code so it is always safe to retry [ExternalError]s, but SQL txns
+    /// will have to pass the error back to the user (or risk double committing
+    /// the txn).
+    ///
+    /// TODO: This already retries [mz_persist::location::Determinate] errors,
+    /// so the signature could be changed to only return Indeterminate, but
+    /// leaving it as ExternalError for now to save churn on storage PR rebases.
+    pub async fn compare_and_append_core(
```

Comment:
I think `append_core` and `compare_and_append_core` should be methods on Batch! that feels pretty natural API wise and it neatly solves the naming problem

we could avoid the need to clone Machine, etc by giving Batch a lifetime and having it mutably borrow `&mut WriteHandle`. it seems fine to me to only allow a WriteHandle to be in the process of constructing one batch at a time

Reply:
I moved these methods to `Batch`, but requiring `Batch` to have a reference to `WriteHandle` felt a bit unnatural, mostly because a successful append changes the cached upper of the `WriteHandle`, so `Batch` has to "reach" into the `WriteHandle` to do that.

I played around with the `concurrency()` test, and that seemed easier with `Batch` not referencing a write handle, though that could be a function of how I implemented it. The last commit moves the `_core` methods back to `WriteHandle` and changes the test, so you can look at it in isolation. At this point it's probably easier to figure out the fine details in a call, but I think I have a good understanding of the ergonomics now.

Reply:
> as for naming, let's call this compare_and_append_batch (ditto append_batch)

will do!